server/service/src/sync/remote_data_synchroniser.rs

```diff
@@ -131,117 +122,129 @@ impl RemoteDataSynchroniser {
     }
 
     // Push all records in change log to central server
-    pub async fn push(&self, connection: &StorageConnection) -> Result<(), anyhow::Error> {
+    pub(crate) async fn push(&self, connection: &StorageConnection) -> Result<(), anyhow::Error> {
         let changelog = ChangelogRowRepository::new(connection);
 
         const BATCH_SIZE: u32 = 1000;
         let state = RemoteSyncState::new(connection);
         loop {
+            // TODO inside transaction
             info!("Remote push: Check changelog...");
             let cursor = state.get_push_cursor()?;
             let changelogs = changelog.changelogs(cursor as u64, BATCH_SIZE)?;
-            if changelogs.is_empty() {
-                break;
-            }
+            let change_logs_total = changelog.count(cursor as u64)?;
+            let total_remaining_after_push = change_logs_total - changelogs.len() as u64;
+            let last_pushed_cursor = changelogs.last().map(|log| log.id);
+
             info!(
                 "Remote push: Translate {} changelogs to push records...",
                 changelogs.len()
             );
-            let last_changelog_id = changelogs.last().map(|log| log.id).unwrap_or(cursor as i64);
+
             let records = translate_changelogs_to_push_records(connection, changelogs)?;
+            let records_len = records.len();
 
-            self.sync_api_v3
-                .post_queued_records(self.site_id, self.central_server_site_id, &records)
+            let response = self
+                .sync_api_v5
+                .post_queued_records(total_remaining_after_push, records)
                 .await?;
 
-            state.update_push_cursor(last_changelog_id as u32 + 1)?;
             info!(
                 "Remote push: {} records pushed to central server",
-                records.len()
+                records_len
             );
+
+            // Update cursor only if record for that cursor has been pushed/processed
+            if let Some(last_pushed_cursor_id) = last_pushed_cursor {
+                state.update_push_cursor(last_pushed_cursor_id as u32 + 1)?;
```

Comment:
have to be careful not to set the cursor + 1 if no records were sent, and we do need to send a push with 0 queuedRecords every time for central server to integrate